### PR TITLE
SceneContentAdjuster updates

### PR DESCRIPTION
**Pull request by keveleigh**
Overview
---
Scene content adjuster both wasn't working on an actual device and was overwriting the X and Z values of the GameObject's position. This fixes both of those bugs.

Changes
---
- Fixes: #1324
- Fixes: #1354 
- Fixes: #1304 
- Fixes: #1384

### CODE REVIEW

File: Assets/HoloToolkit-Examples/Input/Scenes/InputManagerTest.unity [hunk 1]
```diff
@@ -1415,7 +1415,7 @@ RectTransform:
   m_PrefabInternal: {fileID: 0}
   m_GameObject: {fileID: 432117531}
   m_LocalRotation: {x: -0, y: -0, z: -0, w: 1}
-  m_LocalPosition: {x: 0.049999237, y: 0.47200012, z: 1.1980001}
+  m_LocalPosition: {x: 0, y: 0, z: 1.1980001}
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/Input/Scenes/InputManagerTest.unity]: Should Z = 0 as well?

File: Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs [hunk 2]
```diff
@@ -1,46 +1,46 @@
...
+            if (frameWaitHack < 1)
             {
-                lastFloorHeight.y = floorHeightOffset;
-                transform.position = lastFloorHeight;
+                // Not waiting a frame often caused the camera's position to be incorrect at this point. This seems like a Unity bug.
+                frameWaitHack++;
+                yield return null;
             }
+
+            transform.position = new Vector3(transform.position.x, CameraCache.Main.transform.position.y, transform.position.z);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs]: cache new Vector3 and just reset vector to `Vector3.zero` then update values

File: Assets/HoloToolkit-Examples/Input/Scenes/InputManagerTest.unity (hunk 1 above)
Comment (keveleigh) [reply Assets/HoloToolkit-Examples/Input/Scenes/InputManagerTest.unity]: The editor updated this every time I dismissed it. I'll have to look back and verify what Z should be.

File: Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs [hunk 3]
```diff
@@ -1,46 +1,58 @@
...
 
 #if UNITY_2017_2_OR_NEWER
+using HoloToolkit.Unity.Boundary;
 using UnityEngine.XR;
-#else
-using UnityEngine.VR;
 #endif
 
-public class SceneContentAdjuster : MonoBehaviour
+namespace HoloToolkit.Unity
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs]: `HoloToolkit.Unity.Boundary`

File: Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs [hunk 4]
```diff
@@ -28,8 +40,7 @@ private void Awake()
             }
 
 #if UNITY_2017_2_OR_NEWER
-            // A Stationary TrackingSpaceType doesn't need any changes for an object.
-            if (XRDevice.GetTrackingSpaceType() == TrackingSpaceType.Stationary || !XRDevice.isPresent)
+            if (!XRDevice.isPresent)
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs]: Should we be checking is the headset is opaque?

File: Assets/HoloToolkit-Examples/Boundary/Scripts/BoundaryTest.cs [hunk 5]
```diff
@@ -28,6 +28,17 @@ private void Start()
                     defaultMaterials[i].SetColor("_Color", highlightColor);
                 }
             }
+            else
+            {
+                Debug.LogFormat("Object " + name + " is outside established boundary. Position: {0}", gameObject.transform.position);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/Boundary/Scripts/BoundaryTest.cs]: Should this be a warning?

Comment (keveleigh) [reply Assets/HoloToolkit-Examples/Boundary/Scripts/BoundaryTest.cs]: It's more of an informative message in the test scene, showing which of the pre-existing cubes are inside/outside the boundary.

File: Assets/HoloToolkit-Examples/Boundary/Materials/Highlight.mat [hunk 6]
```diff
@@ -7,8 +7,8 @@ Material:
   m_PrefabParentObject: {fileID: 0}
   m_PrefabInternal: {fileID: 0}
   m_Name: Highlight
-  m_Shader: {fileID: 4800000, guid: f326278a838e87c499393f30a8e72127, type: 3}
-  m_ShaderKeywords: ETC1_EXTERNAL_ALPHA _EMISSION
+  m_Shader: {fileID: 4800000, guid: 06790e211e4a6df409b1e04bc91c4810, type: 3}
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/Boundary/Materials/Highlight.mat]: What's the new shader we're using?

File: Assets/HoloToolkit-Examples/Boundary/Scripts/BoundaryTest.cs [hunk 7]
```diff
@@ -19,7 +19,7 @@ private void Start()
         
             if (BoundaryManager.Instance.ContainsObject(gameObject.transform.position))
             {
-                Debug.LogFormat("Object is within established boundary. Position: {0}", gameObject.transform.position);
+                Debug.LogFormat("Object " + name + " is within established boundary. Position: {0}", gameObject.transform.position);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/Boundary/Scripts/BoundaryTest.cs]: `Debug.LogFormat("Object {0} is within established boundary. Position: {1}", name, gameObject.transform.position);

File: Assets/HoloToolkit-Examples/Boundary/Scripts/BoundaryTest.cs (hunk 5 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit-Examples/Boundary/Scripts/BoundaryTest.cs]:
```
Debug.LogFormat("Object {0} is outside established boundary. Position: {1}", name, gameObject.transform.position);
```

File: Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs (hunk 4 above)
Comment (keveleigh) [reply Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs]: Hmm...I don't think so. Although the floor is only set in opaque headsets right now, that doesn't necessarily have to be the case forever. The height offset really only comes from Stationary vs RoomScale.

File: Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs [hunk 8]
```diff
@@ -104,45 +116,18 @@ protected override void Awake()
...
-            if (FloorQuad != null && HolographicSettings.IsDisplayOpaque)
+            if (FloorQuad != null && XRDevice.GetTrackingSpaceType() == TrackingSpaceType.RoomScale)
             {
                 floorQuadInstance = Instantiate(FloorQuad);
 
                 if (!XRDevice.isPresent)
                 {
                     // So the floor quad does not occlude in editor testing, draw it lower.
-                    floorQuadInstance.transform.localPosition = new Vector3(0, -3, 0);
+                    floorQuadInstance.transform.position = new Vector3(0, -3, 0);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: Should this be cached as a const?

File: Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs [hunk 9]
```diff
@@ -7,40 +10,73 @@
...
+            }
 
-            if (lastFloorHeight.y != floorHeightOffset)
+            if (XRDevice.GetTrackingSpaceType() == TrackingSpaceType.RoomScale)
             {
-                lastFloorHeight.y = floorHeightOffset;
-                transform.position = lastFloorHeight;
+                if (alignWithInitialHeadHeight)
+                {
+                    containerObject.position = new Vector3(containerObject.position.x, containerObject.position.y + CameraCache.Main.transform.position.y, containerObject.position.z);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs]: We could probably use a local Vector3 field and just reset it, then update it with the new position before assigning the container objects new position

File: Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs (hunk 4 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs]: Could we add a note about that here please?

File: Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs (hunk 9 above)
Comment (keveleigh) [reply Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs]: Would there be a benefit to that change?

Comment (StephenHodgson) [reply Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs]: Removes the need for majick numbers and reduces the need to allocate a new Vector3

Comment (keveleigh) [reply Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs]: Hmm true. As-is, this method is only called once, but I'll update this to future-proof it a bit.

File: Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs (hunk 8 above)
Comment (keveleigh) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: image

File: Assets/HoloToolkit-Examples/Boundary/Materials/Highlight.mat (hunk 6 above)
Comment (keveleigh) [reply Assets/HoloToolkit-Examples/Boundary/Materials/Highlight.mat]: Unlit Configurable. The old guid is missing, and the material had the magenta error shader applied.

File: Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs (hunk 8 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: Ah not as a const but a readonly private field.

Comment (StephenHodgson) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: Why are we choosing `-3` anyway?

Comment (keveleigh) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: > Why are we choosing -3 anyway?

Experimentation in the editor found that this got it out of the way, and is only used in the case where a headset isn't attached. Otherwise, it functions as normal.

It might not even be worth showing the floor in the editor at an arbitrary location, since it doesn't mean much without a headset attached and a boundary defined.

Comment (StephenHodgson) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: Should we add these comments to our description of the value?

**Comment by jessemcculloch** [comment]
The boundaryTest scene is throwing the following error when I run it in the editor.  Not sure if this is something unrelated to this PR or not.

transform.localPosition assign attempt for 'LeftController' is not valid. Input localPosition is { 0.000000, NaN, 0.000000 }.
UnityEngine.Transform:set_localPosition(Vector3)
HoloToolkit.Unity.InputModule.MotionControllerVisualizer:UpdateControllerState() (at Assets/HoloToolkit/Input/Scripts/Utilities/MotionControllerVisualizer.cs:162)
HoloToolkit.Unity.InputModule.MotionControllerVisualizer:Application_onBeforeRender() (at Assets/HoloToolkit/Input/Scripts/Utilities/MotionControllerVisualizer.cs:123)

**Comment by keveleigh** [comment]
@jessemcculloch Unrelated to this PR, but something we (or the platform) should address. See #1290

File: Assets/HoloToolkit-Examples/UX/Scripts/OptimizeSceneforDeviceType.cs [hunk 10]
```diff
@@ -2,35 +2,15 @@
...
+using System;
 using UnityEngine;
-using UnityEngine.XR.WSA;
 
 namespace HoloToolkit.Unity.Examples
 {
+    [Obsolete("Use SceneContentAdjuster")]
     public class OptimizeSceneforDeviceType : MonoBehaviour
     {
         public GameObject containerObject;
```
Comment (NeerajW) [inline Assets/HoloToolkit-Examples/UX/Scripts/OptimizeSceneforDeviceType.cs]: Why not delete public GameObject as well?

Comment (keveleigh) [reply Assets/HoloToolkit-Examples/UX/Scripts/OptimizeSceneforDeviceType.cs]: To prevent broken links in the Inspector, to help make porting to the new script as easy as possible.

File: Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs [hunk 11]
```diff
@@ -7,40 +10,82 @@
...
 #endif
 
-public class SceneContentAdjuster : MonoBehaviour
+namespace HoloToolkit.Unity.Boundary
 {
-    private Vector3 lastFloorHeight;
-    private float floorHeightOffset = 1f;
-
-    private void Awake()
+    public class SceneContentAdjuster : MonoBehaviour
```
Comment (NeerajW) [inline Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs]: Pls add summary. What does this class actually do?

File: Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs [hunk 12]
```diff
@@ -20,23 +20,34 @@ public class BoundaryManager : Singleton<BoundaryManager>
...
         [SerializeField]
         [Tooltip("Approximate max Y height of your space.")]
         private float boundaryHeight = 10f;
+
         private Bounds boundaryBounds;
+
+        [SerializeField]
+        // Defaulting coordinate system to RoomScale in immersive headsets.
+        // This puts the origin (0, 0, 0) on the floor if a floor has been established during setup via MixedRealityPortal.
+        private TrackingSpaceType opaqueTrackingSpaceType = TrackingSpaceType.RoomScale;
```
Comment (NeerajW) [inline Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: These made the code hard to read but okay.

Comment (keveleigh) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: This allows the dev to select which tracking space type they want, to more easily support seated-experience apps like MixedReality250.

File: Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs [hunk 13]
```diff
@@ -7,40 +10,82 @@
...
+        [Tooltip("Optional container object reference. If null, this script will move the object it's attached to.")]
+        private Transform containerObject;
+
+        [SerializeField]
+        [Tooltip("Select this if the container should be placed in front of the head on app launch in a room scale app.")]
+        private bool alignWithHeadHeight = true;
+
+        [SerializeField]
+        [Tooltip("Use this to set the desired position of the container in a stationary app.")]
+        private Vector3 stationarySpaceTypePosition = Vector3.zero;
```
Comment (NeerajW) [inline Assets/HoloToolkit/Boundary/Scripts/SceneContentAdjuster.cs]: These I think will cause more confusion than solve.
Perhaps more text around how or when to tweak them would be good.

File: Assets/HoloToolkit/Input/Prefabs/MixedRealityCameraParent.prefab [hunk 14]
```diff
@@ -667,6 +667,8 @@ MonoBehaviour:
   m_EditorClassIdentifier: 
   FloorQuad: {fileID: 1520281459093612, guid: 08d25ccfbbc595b4a820c7edd551d095, type: 2}
   boundaryHeight: 10
+  opaqueTrackingSpaceType: 1
```
Comment (NeerajW) [inline Assets/HoloToolkit/Input/Prefabs/MixedRealityCameraParent.prefab]: Why don't we just set it in MixedRealityCameraManager then? Instead of BoundaryManager?

Comment (keveleigh) [reply Assets/HoloToolkit/Input/Prefabs/MixedRealityCameraParent.prefab]: TrackingSpaceType isn't a camera setting, and boundary and tracking space type are strongly linked (and this was already being set in BoundaryManager; it just wasn't configurable until this change).

File: Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs (hunk 12 above)
Comment (NeerajW) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: MR250 still sets it as full scale  for immersive.

File: Assets/HoloToolkit/Input/Prefabs/MixedRealityCameraParent.prefab (hunk 14 above)
Comment (NeerajW) [reply Assets/HoloToolkit/Input/Prefabs/MixedRealityCameraParent.prefab]: How can tracking not be related to camera?

File: Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs [hunk 15]
```diff
@@ -172,7 +156,7 @@ public bool ContainsObject(Vector3 gameObjectPosition)
         public void CalculateBoundaryVolume()
         {
             // TODO: BUG: Unity: Should return true if a floor and boundary has been established by user.
-            // But this always returns false with in 2017.2.0f3 MRTP3.
+            // But this always returns false with in 2017.2.0p1-MRTP4.
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: Should we update this to MRTP 5?

File: Assets/HoloToolkit/UX/Scripts/Receivers/InteractionReceiver.cs [hunk 16]
```diff
@@ -70,8 +70,15 @@ public virtual void OnEnable()
...
-            InputManager.Instance.RemoveGlobalListener(gameObject);
-            FocusManager.Instance.PointerSpecificFocusChanged += OnPointerSpecificFocusChanged;
+            if (InputManager.IsInitialized)
+            {
+                InputManager.Instance.RemoveGlobalListener(gameObject);
+            }
+
+            if (FocusManager.IsInitialized)
+            {
+                FocusManager.Instance.PointerSpecificFocusChanged -= OnPointerSpecificFocusChanged;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UX/Scripts/Receivers/InteractionReceiver.cs]: Isn't this an event in the InputManager?

We probably don't need to subscribe to this kind of event if we can get it through the interface.  Might see a perf improvement here.

File: Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs (hunk 15 above)
Comment (keveleigh) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: I'll take a look.

File: Assets/HoloToolkit/UX/Scripts/Receivers/InteractionReceiver.cs (hunk 16 above)
Comment (keveleigh) [reply Assets/HoloToolkit/UX/Scripts/Receivers/InteractionReceiver.cs]: My change was wrapping the lines in the `IsInitialized` checks to prevent null refs on teardown.

That sounds like something to look into though. Perhaps https://github.com/Microsoft/MixedRealityToolkit-Unity/blob/master/Assets/HoloToolkit/Input/Scripts/InputHandlers/IPointerSpecificFocusable.cs

Comment (StephenHodgson) [reply Assets/HoloToolkit/UX/Scripts/Receivers/InteractionReceiver.cs]: Exactly the interface I was talking about :)

File: Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs [hunk 17]
```diff
@@ -181,13 +165,16 @@ public void CalculateBoundaryVolume()
...
                 return;
             }
 
             boundaryBounds = new Bounds();
+
             // Get all the bounds setup by the user.
             var boundaryGeometry = new List<Vector3>(0);
+            // TODO: BUG: Unity: Should return true if a floor and boundary has been established by user.
+            // But this always returns false with in 2017.2.0p2-MRTP5.
             if (UnityEngine.Experimental.XR.Boundary.TryGetGeometry(boundaryGeometry))
```
Comment (NeerajW) [inline Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: Worth adding an else block and dumping a debug log.
